fl4health/model_bases/masked_layers/masked_linear.py

```diff
@@ -0,0 +1,92 @@
+from typing import Optional
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch import Tensor
+from torch.nn.parameter import Parameter
+
+from fl4health.utils.functions import bernoulli_sample
+
+
+class MaskedLinear(nn.Linear):
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        bias: bool = True,
+        device: Optional[torch.device] = None,
+        dtype: Optional[torch.dtype] = None,
+    ) -> None:
+        """
+        Implementation of masked linear layers.
+
+        Like regular linear layers (i.e., nn.Linear module), a masked linear layer has a weight and a bias.
+        However, the weight and the bias do not receive gradient in back propagation.
+        Instead, two score tensors - one for the weight and another for the bias - are maintained.
+        In the forward pass, the score tensors are transformed by the Sigmoid function into probability scores,
+        which are then used to produce binary masks via bernoulli sampling.
+        Finally, the binary masks are applied to the weight and the bias. During training,
+        gradients with respect to the score tensors are computed and used to update the score tensors.
+
+        Note: the scores are not assumed to be bounded between 0 and 1.
+
+        Args:
+            in_features: size of each input sample
+            out_features: size of each output sample
+            bias: If set to ``False``, the layer will not learn an additive bias.
+                Default: ``True``
+
+        Attributes:
+            weight: weights of the module.
+            bias:  bias of the module.
+            weight_score: learnable scores for the weights. Has the same shape as weight.
+            bias_score: learnable scores for the bias. Has the same shape as bias.
+        """
+        super().__init__(in_features, out_features, bias, device, dtype)
+        self.in_features = in_features
+        self.out_features = out_features
+        self.weight.requires_grad = False
+        self.weight_scores = Parameter(torch.randn_like(self.weight), requires_grad=True)
+        if bias:
+            assert self.bias is not None
+            self.bias.requires_grad = False
+            self.bias_scores = Parameter(torch.randn_like(self.bias), requires_grad=True)
+        else:
+            self.register_parameter("bias", None)
```

Comment:
If bias is None, do we need to register it as such here? For example, we don't do this for the normalization layers below.

Reply:
You are right. It is already registered as None in the parent class `nn.Linear`.